File: Ruby/stack.rb


class Stack

	def initialize (size)
		@size_ = size
		@top_ = -1
		@stack_ = Array.new(@size_)
	end

	def push (dato)
		if !full
			@top_+=1
			@stack_[@top_] = dato
		else
			abort "Pila llena!"
		end
	end

	def pop
		if !empty
			dato = @stack_[@top_]
			@top_-=1
			dato
		else
			abort "Pila vacia!"
		end
	end

	def top
		if !empty
			@stack_[top_]
		else
			abort "Pila vacia!"
		end
	end

	def empty
		@top_ == -1
	end

	def full
		@top_ >= @size_ - 1
	end

	def show
		for x in (0..@top_)
			print "#{@stack_[x]}, "
		end
		puts
	end
end

pila = Stack.new(10)
pila.push 'a'
pila.push 'b'
pila.push 2.4
pila.push 6
pila.show
pila.pop
pila.pop
pila.push 'z'
pila.show

File: C++/stack.cpp
#include <iostream>
#include <cstdlib>

using namespace std;


class Stack {

	private:

		int* stack_;
		int size_;
		int top_;

	public:

		Stack(int size):
			size_(size),
			top_(-1),
			stack_(NULL){
				stack_ = new int [size_];
			}

		~Stack(){

			if (stack_){
				delete stack_;
				stack_ = NULL;
			}
		}

		void push(int dato){

			if (!full()){
				top_++;
				stack_[top_] = dato;
			}

			else
				cout << "Pila llena!" << endl;
		}

		int pop(){

			if (!empty()){
				int dato = stack_[top_];
				top_--;
				return dato;
			}

			else{
				cout << "Pila vacia!" << endl;
				exit(1);		
			}
		}

		bool empty(){

			return (top_ == -1);
		}

		bool full(){

			return (top_ >= size_-1);
		}

		void show(){

			for (int i=0; i<=top_; i++)
				cout << stack_[i] << ", ";
		}

		int top(){
			if (!empty())
				return stack_[top_];

			else{
				cout << "Pila vacia!" << endl;
				exit(1);		
			}			

		}
};



int main(void){
	

	Stack Pila(10);

	cout << "Mostrar pila: "; Pila.show(); cout << endl;
	Pila.push(1);
	Pila.push(5);
	Pila.push(4);
	Pila.push(2);
	cout << "Mostrar pila: "; Pila.show(); cout << endl;
	int pop = Pila.pop();
	pop = Pila.pop();
	Pila.push(9);
	cout << "Mostrar pila: "; Pila.show(); cout << endl;

	return 0;
}